File: src/p0003/erlang/solution.erl
-spec length_of_longest_substring(S :: unicode:unicode_binary()) -> integer().
length_of_longest_substring(S) ->
    L = erlang:binary_to_list(S),
    loop_j(0, L, 0, L, array:new(256, {default, 0}), 0).

loop_j(_, _, _, [], _, Result) -> Result;
loop_j(I, LI, J, [Y | _] = LJ, Count, Result) ->
    loop_i(I, LI, J, LJ, array_incr(Y, 1, Count), Result).

loop_i(I, [X | TI] = LI, J, [Y | TJ] = LJ, Count, Result) ->
    case array:get(Y, Count) > 1 of
        true ->
            loop_i(I + 1, TI, J, LJ, array_incr(X, -1, Count), Result);
        false ->
            loop_j(I, LI, J + 1, TJ, Count, max(Result, J - I + 1))
    end.

array_incr(I, D, Array) -> array:set(I, array:get(I, Array) + D, Array).

File: src/p0003/elixir/solution.ex
defmodule Solution do
  @spec length_of_longest_substring(s :: String.t) :: integer
  def length_of_longest_substring(s) do
    l = String.to_charlist(s)
    loop_j(0, l, 0, l, :array.new(256, {:default, 0}), 0)
  end

  def loop_j(_, _, _, [], _, result), do: result
  def loop_j(i, li, j, [y | _] = lj, count, result) do
    loop_i(i, li, j, lj, array_incr(y, 1, count), result)
  end

  def loop_i(i, [x | ti] = li, j, [y | tj] = lj, count, result) do
    case :array.get(y, count) > 1 do
      true ->
        loop_i(i + 1, ti, j, lj, array_incr(x, -1, count), result)
      _ ->
        loop_j(i, li, j + 1, tj, count, max(result, j - i + 1))
    end
  end

  def array_incr(i, d, array) do
    :array.set(i, :array.get(i, array) + d, array)
  end
end
